components/script/canvas_state.rs

```diff
@@ -60,6 +60,39 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
 use crate::dom::textmetrics::TextMetrics;
 use crate::script_runtime::CanGc;
 
+// Firefox limits width/height to 32767 pixels and Chromium to 65535 pixels,
+// but slows down dramatically before it reaches that limit.
+// We limit by area instead, giving us larger maximum dimensions,
+// in exchange for a smaller maximum canvas size.
+const MAX_CANVAS_AREA: u64 = 32768 * 8192;
+// Max width/height to 65535 in CSS pixels.
+const MAX_CANVAS_SIZE: u64 = 65535;
+
+#[derive(Clone, Debug, JSTraceable, MallocSizeOf)]
+/// Helps observe states of canvas draw target.
+struct DrawTarget {
```

Comment:
The use of `DrawTarget` here is a little bit confusing, because this isn't the target of the draw itself, but metadata about it. I would call it `DrawTargetSize`.

```suggestion
struct DrawTargetSize {
```

Reply:
Yes, it has mostly metadata in itself, but the goal is to have mirrored version (with size + state) of canvas draw target but on script thread. Later I have plan to add delayed creation of draw target for Canvas2D context (was cutoff from current version of PR) and it will add new filed "created/allocated" to this struct.

I was inspired by existed WebGPU implementation in Servo - it also have the same thing - metadata version of WebGPU drawing buffer on script thread.
https://github.com/servo/servo/blob/main/components/script/dom/webgpu/gpucanvascontext.rs#L54

Canvas2D [output bitmap](https://html.spec.whatwg.org/multipage/canvas.html#2dcontext:output-bitmap) -> CanvasState::DrawTarget
WebGPU [drawing buffer](https://gpuweb.github.io/gpuweb/#dom-gpucanvascontext-drawingbuffer-slot) -> GPUCanvasContext:DrawingBuffer
WebGL [drawing buffer](https://registry.khronos.org/webgl/specs/latest/1.0/#2.2) -> None

Anyway if after you and/or @sagudev decide to rename it to DrawTargetSize - no problem.

Reply:
Okay. That plan sounds interesting. For now let's give it a name that matches what it does. It's no problem to rename it later.

Reply:
> I was inspired by existed WebGPU implementation in Servo - it also have the same thing - metadata version of WebGPU drawing buffer on script thread.

The reason why we named it this way in WebGPU is because it matches the spec and it's also used referenced in spec steps (this way also out code matches spec better).

But here there are no exact steps written out in spec so either name is fine by me.

Reply:
Okay. Let's call this `DrawTargetSize` then. Thanks!